contracts/Redemption.vy

```diff
@@ -0,0 +1,489 @@
+    eth_per_yfi: uint256 = amount * self._get_latest_price() / UNIT
+    return eth_per_yfi * (UNIT - self._discount()) / UNIT
+
+
+@external
+@view
+def get_latest_price() -> uint256:
+    """
+    @notice Get the latest price of YFI in ETH
+    @return Price of YFI in ETH (18 decimals)
+    """
+    return self._get_latest_price()
+
+
+@internal
+@view
+def _get_latest_price() -> uint256:
+    oracle_price: uint256 = convert(self._get_oracle_price(), uint256)
+    pool_price: uint256 = CURVE_POOL.price_oracle()
+    if pool_price < oracle_price:
+        return oracle_price
+    return pool_price
+
+
+@internal
+@view
+def _get_oracle_price() -> int256:
+    round_id: uint80 = 0
+    price: int256 = 0
+    started_at: uint256 = 0
+    updated_at: uint256 = 0
+    answered_in_round: uint80 = 0
+    (round_id, price, started_at, updated_at, answered_in_round) = PRICE_FEED.latestRoundData()
+    assert updated_at + 86400 > block.timestamp, "price too old"
+    return price
+
+
+@external
+@view
+def scaling_factor() -> uint256:
+    """
+    @notice Get the current discount curve scaling factor
+    @return Scaling factor (18 decimals)
+    """
+    return convert(self._scaling_factor()[0], uint256)
+
+@external
+@view
+def scaling_factor_ramp() -> (uint256, uint256, uint256, uint256):
+    """
+    @notice Get the current discount curve scaling factor ramp parameters
+    @return Tuple of ramp start timestamp, ramp end timestamp, ramp start scaling factor, ramp end scaling factor
+    """
+    ramp_start: uint256 = 0
+    ramp_end: uint256 = 0
+    old: int256 = 0
+    new: int256 = 0
+    ramp_start, ramp_end, old, new = self._unpack_scaling_factor(self.packed_scaling_factor)
+    return ramp_start, ramp_end, convert(old, uint256), convert(new, uint256)
+
+
+@internal
+@view
+def _scaling_factor() -> (int256, bool):
+    ramp_start: uint256 = 0
+    ramp_end: uint256 = 0
+    old: int256 = 0
+    new: int256 = 0
+    ramp_start, ramp_end, old, new = self._unpack_scaling_factor(self.packed_scaling_factor)
+    if ramp_end <= block.timestamp:
+        return new, False
+    if ramp_start > block.timestamp:
+        return old, False
+    
+    duration: int256 = convert(ramp_end - ramp_start, int256)
+    time: int256 = convert(block.timestamp - ramp_start, int256)
+    return old + (new - old) * time / duration, True
+
+
+@external
+def set_payee(new_payee: address):
+    """
+    @dev set the payee of the ETH used for redemptions
+    @param new_payee the new payee
+    """
+    self._check_owner()
+    self.payee = new_payee
+    log SetPayee(new_payee)
+
+
+@external
+def start_ramp(new: uint256, duration: uint256 = 604_800, start: uint256 = block.timestamp):
+    """
+    @notice Start ramping of scaling factor
+    @param new New scaling factor (18 decimals)
+    @param duration Ramp duration (seconds)
+    @param start Ramp start timestamp
+    """
+    self._check_owner()
+    assert new <= 12 * UNIT
```

Comment:
it looks like a `new` of 0 would give it a 100% discount and free YFI.

Perhaps another place to add a sanity check. Looks like the YIP specifies it must be >= 1.

Reply:
i guess the new > 0 check is in the _pack_scaling_factor function but still may make sense to enforce the yip spec of >= 1